File: go-backend/internals/services/service.go
package services

import (
	"context"
	"quikbites-be/internals/models"
)

type UserService interface {
	Register(ctx context.Context, user models.User) error
	Login(ctx context.Context, credentials models.User) (models.LoginResponse, error)
	GetProfile(ctx context.Context, userId int) (models.User, error)
	UpdateProfile(ctx context.Context, user models.User) error
	UpdateCurrentAddress(ctx context.Context, userId int, address models.Address) error
}

type RestaurantService interface {
	GetRestaurants(ctx context.Context, lat, long float64, query string) ([]models.Restaurant, error)
	GetRestaurantDetails(ctx context.Context, restaurantID int) (models.RestaurantDetails, error)
	UpdateRestaurantDetails(ctx context.Context, restaurant models.Restaurant) error
	ToggleAvailability(ctx context.Context, restaurantID int, isOpen bool) error

	GetMenu(ctx context.Context, restaurantID int) ([]models.MenuItem, error)
	AddMenuItem(ctx context.Context, restaurantID int, item models.MenuItem) error
	UpdateMenuItem(ctx context.Context, item models.MenuItem) error
	DeleteMenuItem(ctx context.Context, itemId int) error

	GetRestaurantOrders(ctx context.Context, restaurantID int) ([]models.OrderDetail, error)
	UpdateOrderStatus(ctx context.Context, orderId int, status string) error
	GetAnalytics(ctx context.Context, restaurantID int) (models.RestaurantAnalytics, error)
}

type CartService interface {
	GetCart(ctx context.Context, userId, restaurantID int) ([]models.CartItem, error)
	AddItemToCart(ctx context.Context, userId, restaurantID, itemId int) error
	RemoveItemFromCart(ctx context.Context, userId, restaurantID, itemId int) error
	ClearCart(ctx context.Context, userId, restaurantID int) error
}

type OrderService interface {
	PlaceOrder(ctx context.Context, userId, restaurantID int, paymentMode string) (models.OrderDetail, error)
	GetOrderHistory(ctx context.Context, userId int) ([]models.OrderDetail, error)
	GetActiveOrders(ctx context.Context, userId int) ([]models.OrderDetail, error)
	TrackOrder(ctx context.Context, userId, orderId int) (models.OrderDetail, error)
	VerifyDeliveryOTP(ctx context.Context, userId, orderId int, otp string) error
	ContactDriver(ctx context.Context, orderId int) (string, error)
}

type DeliveryService interface {
	GetAvailableOrders(ctx context.Context, driverID int) ([]models.OrderDetail, error)
	AcceptOrder(ctx context.Context, driverID, orderID int) error
	GetActiveOrders(ctx context.Context, driverID int) ([]models.OrderDetail, error)
	UpdateOrderStatus(ctx context.Context, driverID, orderID int, status string) error
	VerifyDeliveryCompletion(ctx context.Context, driverID, orderID int, otp string) error
	GetOrderHistory(ctx context.Context, driverID int) ([]models.OrderDetail, error)
	GetAnalytics(ctx context.Context, driverID int) (models.DriverAnalytics, error)
	ContactCustomer(ctx context.Context, orderID int) (string, error)
	UpdateLocation(ctx context.Context, driverID int, lat, long float64) error
}

File: express-backend/service/service.ts
import {
    Address,
    CartItem,
    DriverAnalytics,
    LoginResponse,
    MenuItem,
    OrderDetail,
    Restaurant,
    RestaurantAnalytics,
    RestaurantDetails,
    User,
} from '../model/model';

export interface UserService {
    register(user: User): Promise<void>;
    login(credentials: User): Promise<LoginResponse>;
    getProfile(userId: number): Promise<User>;
    updateProfile(user: User): Promise<void>;
    updateCurrentAddress(userId: number, address: Address): Promise<void>;
}

export interface RestaurantService {
    getRestaurants(lat: number, long: number, query: string): Promise<Restaurant[]>;
    getRestaurantDetails(restaurantID: number): Promise<RestaurantDetails>;
    updateRestaurantDetails(restaurant: Restaurant): Promise<void>;
    toggleAvailability(restaurantID: number, isOpen: boolean): Promise<void>;

    getMenu(restaurantID: number): Promise<MenuItem[]>;
    addMenuItem(restaurantID: number, item: MenuItem): Promise<void>;
    updateMenuItem(item: MenuItem): Promise<void>;
    deleteMenuItem(itemId: number): Promise<void>;

    getRestaurantOrders(restaurantID: number): Promise<OrderDetail[]>;
    updateOrderStatus(orderId: number, status: string): Promise<void>;
    getAnalytics(restaurantID: number): Promise<RestaurantAnalytics>;
}

export interface CartService {
    getCart(userId: number, restaurantID: number): Promise<CartItem[]>;
    addItemToCart(userId: number, restaurantID: number, itemId: number): Promise<void>;
    removeItemFromCart(userId: number, restaurantID: number, itemId: number): Promise<void>;
    clearCart(userId: number, restaurantID: number): Promise<void>;
}

export interface OrderService {
    placeOrder(userId: number, restaurantID: number, paymentMode: string): Promise<OrderDetail>;
    getOrderHistory(userId: number): Promise<OrderDetail[]>;
    getActiveOrders(userId: number): Promise<OrderDetail[]>;
    trackOrder(userId: number, orderId: number): Promise<OrderDetail>;
    verifyDeliveryOTP(userId: number, orderId: number, otp: string): Promise<void>;
    contactDriver(orderId: number): Promise<string>;
}

export interface DeliveryService {
    getAvailableOrders(driverID: number): Promise<OrderDetail[]>;
    acceptOrder(driverID: number, orderID: number): Promise<void>;
    getActiveOrders(driverID: number): Promise<OrderDetail[]>;
    updateOrderStatus(driverID: number, orderID: number, status: string): Promise<void>;
    verifyDeliveryCompletion(driverID: number, orderID: number, otp: string): Promise<void>;
    getOrderHistory(driverID: number): Promise<OrderDetail[]>;
    getAnalytics(driverID: number): Promise<DriverAnalytics>;
    contactCustomer(orderID: number): Promise<string>;
    updateLocation(driverID: number, lat: number, long: number): Promise<void>;
}



